File: Linked_List/Reverse_a_linked_list_in_groups_of_a_given_size/java/src/MyLinkedList.java
public class MyLinkedList<T> {
    Node<T> head = null;

    public Node<T> reverse(Node<T> head, int k) {
        Node<T> current = head;
        Node<T> next = null;
        Node<T> previous = null;
        int count = 0;

        while (current != null && count < k) {
            next = current.next;
            current.next = previous;
            previous = current;
            current = next;
            count++;
        }

        if (next != null) {
            head.next = reverse(next, k);
        }

        return previous;
    }

    public void push(T newData) {
        Node<T> newNode = new Node<>(newData);
        newNode.next = this.head;
        this.head = newNode;
    }

    @Override
    public String toString() {
        Node<T> temp = head;
        StringBuilder builder = new StringBuilder();

        while (temp != null) {
            builder.append(temp.data);
            temp = temp.next;

            if (temp != null) {
                builder.append("--->");
            } else {
                builder.append("--->NULL");
            }
        }

        return builder.toString();
    }
}

File: Linked_List/Reverse_a_linked_list_in_groups_of_a_given_size/kotlin/src/main/kotlin/MyLinkedList.kt
import java.lang.StringBuilder

class MyLinkedList<T> {
    var head: Node<T>? = null

    fun reverse(head: Node<T>?, k: Int): Node<T>? {
        var previous: Node<T>? = null
        var next: Node<T>? = null
        var current = head
        var count = 0

        while (current != null && count < k) {
            next = current.next
            current.next = previous
            previous = current
            current = next
            count++
        }

        if (next != null) {
            head?.next = reverse(next, k)
        }

        return previous
    }

    fun push(newData: T) {
        val newNode = Node(newData)
        newNode.next = head
        head = newNode
    }

    override fun toString(): String {
        var temp = head
        val builder = StringBuilder()
        while (temp != null) {
            builder.append(temp.data)
            temp = temp.next
            if (temp != null) {
                builder.append("--->")
            } else {
                builder.append("--->NULL")
            }
        }
        return builder.toString()
    }
}